### DISCUSSION TITLE
The giant Idea Guy(tm) dumping thread

### DISCUSSION THREAD

**Discussion by Mixarar** [ErinaSugino/Starbound-SBR-Mod#46] [Ideas]
**Description**
Basically after playing for a while I feel like there isn't enough realism regarding to NPCs, and as such it would be cool to see following features:    Small features at the end

- When NPCs are arroused to 100 percent, they start having different dialogues(more suggestive), perhaps following the player or other NPCs trying to persuade them(could cause other NPCs to react to their request and depend on the arrousal either decline or accept the offer) or smth like that.
- In the lustbound outpost thingy, you could make either some sort of interaction with NPCs to ask them out essentially or smth like that, or a much much simpler thing, you could make so that NPCs in Lustbound outpost would have their arrousal gained much faster than the regular world.
- Adding dialogues as a whole would've been a good feature, as currently NPCs are essentially talking without actual speech bubbles appearing. Although this is more like a suggestion to just make a separate mod for dialogues, and isn't really fitting here. (In my dreams it could be like a proper dialogue system where you could be able to select like different things to say like in fallout games or smth)
- Adding more dialogues to the sехtаlk(for vanilla races), and adding a separate placeholder for unsupported races which could be more or less neutral(so that the unsupported races wouldn't use the Human as a placeholder)
- Аnаl/Vаginаl selector(no explanation needed)    EDIT: This would be hard to impement with POV as you would have to also make different a bit new anims
- Abort pill(not sure which mod adds it, i suppose SBR) should work less when further into the pregnancy, and should work more as like a chance based system.
- A pill to stop getting рrеgnаnt, essentially same as prеsеrvаtivеs(So with a chance for it to not work) but for females.
- Different chance to become рrеgnаnt from different races(would be more realistic as in real life the further away the species from another species, the less chance to get prеgnаnt)     EDIT: Possibly also make a config option for a really small chance to get рrеgnаnt from every race
- Possibly make a tool that would allow to configure setting for рrеgnаnciеs and stuff, as playing with over 20 races and making the config for them to be compatible with each other, or smth like that could be quite annoying.
- Return the quests as they were the only way to have some sort of dialogue with an NPC.
- With integration of Lustbound I suppose you wouldn't me suggesting adding an optional posibility to become a whоrе and possibly an optional quest line or smth for that.

p.s. thats all for now as my imagination is almost gone, but I'll probably return to suggest some more stuff.

**Reason**
Well a lot more realism and new possiblities as well as possibly some new кinкs.

**Have you checked existing requests?**
Some of this might've been posted somewhere but hopefully not the whole thing.

**Additional context**
I know that something here wouldn't be added, but still I am looking forward for a response for some atleast small stuff.

**Comment by Mixarar** [comment]


**Reply by ErinaSugino** [reply]
The game tells you quiet obviously in the log when a config cannot be loaded - even what you messed up exactly where. Since failing to load the config will most likely brick the entire action, you will realize something is off, and either read it in the log yourself or toss the log into troubleshooting and we will tell you.

Getting the raceID, however, is nothing I can do about. I cannot read the id of every _installed_ species in-game either. Just fetch it from loaded characters.

**Comment by ErinaSugino** [comment]
Converted into an "Ideas" discussion because this is a lot of ideas all over the place, some of which need discussion.

1. More fluent and versatile NPC AI would certainly be great and is one of my overall goal, but currently NPC behaviour trees are black magic for me, so I cannot give any promises on progress on that one yet.
2. Incredibly high arousal gain in Lustia is a good idea. Have to look into how to best implement a local override of the arousal gain value.
3. This makes no sense. There is dialogue in the mod in the form of the sextalk plugin. Which is displayed as speech bubbles over the actor's head on the node. If this doesn't appear for you, you should troubleshoot that. If you mean stuff outside of a sexnode or your interactive VN dialogue suggestion with answers ala Fallout, this isn't the focus of this mod. For a full interactive dating sim experience, you should look into "Matters of The Heart" - albeit currently paused development, it adds exactly that, and was originally planned to one day have SxB integration.
4. You misunderstand the current sextalk dialogue structure. Unsupported races don't use human dialogue. It just greatly overlaps, since humans use the most generic placeholder dialogue for most part. Writing dialogue isn't my job as a coder - tried it once, almost killed Juni with Cringe - so that would require contributions from others. The original contributors to "Sexbound Extended Dialogues" (which is already integrated) aren't active anymore, though. All I can promise on my part is that SBR 1.2 is dedicated to rewriting the sextalk system to my modular and dynamic.
5. That is a potentially planned idea for later down the road. I know it will mess up POV and create a lot of work for them to update all POV mods, but - progress is progress.
6. I also always forget which items came from which mod - just shows how much everyone used them together, so things should be easier now that they are one. Chance based abortion? Considering it's a pill that might be more realistic, but is it worth the extra trouble for the player? Also, players could just buy a bunch of them and chug them until even a 1% chance triggers, so if one were to stay commited to that idea, huge cooldowns and penalties needed to be applied, and I know how much people _looooove_ being told how to play their game.
7. There is a birth control pill. Again not with a chance, might introduce that with whatever update I'll rework condoms in. Also, all effects currently work regardless of gender. Need to rework that at some point too.
8. Again a feature that makes sense from a logical point of view, but less so from the players perspective. People are already collectively failing to figure out the `compatibleSpecies` table which defines which species can impregnate which species, and I don't know of any race compat mod that actively updated to use the new `speciesType` system. Forcing everyone to define (or undefine to disable the feature) a giant nested list with a shitton of probabilities doesn't sound like a good upgrade.
9. That's what the `speciesType` system was invented for. Every race support creator could add this parameter to their race to assign it a specie type, in which all members are natively compatible with eachother without the need of the config. Hasn't been used by any races so far though. An external tool however is less of a priority. It's technically not part of the mod, as Starbound mods can obviously not edit files (like their own configs) with an in-game menu, and I'll not include any thirdparty executables or stuff in my mod. Also, when it comes to dedicating my time to projects, finishing the animation file editor so that people contributing great stuff who have to work with the over 10k lines long animations file**S** have it easier is much higher priority.
10. That is a plan already, just needs figuring out how to do it without it breaking like it did before - hence they were removed in OG SxB.
11. That could be done in many different ways, and is a complex mechanic needing lots of thought put into it.

<hr>

Additional suggestions:
1. That is an idea fitting for 1.1, I'll keep that in mind after I finished the big features already planned/in the works.
2. Fixing defeat is a goal - for the undefined future.
3. Another viable idea, albeit not high priority.
4. Different stats during sex per race are a planned feature.

**Reply by ErinaSugino** [reply]
That configurability, albeit obviously the right choice for usability of the mod, is a problem in itself. In order for an already existing item to react to config changes, you'd have to query it from the entity via an entity messages. Which means upon right-clicking to use the pill, a message is networked from the client to the world instance server, who looks up the parent entity from the list of all entities, then sends the message probably a tick later. Which then receives the message, fetches the config value, sends it back the exact same route. Two times data has to go through the server with delay for no reason, and get lagged due to bad internet or worst case even lost. All for taking a pill that has a configurable chance to do nothing.

Someone made a mod that's just a pre-filled compatibleSpecies table already - although more aimed at all the Terato monsters than custom races. Other than that, one can always just turn the compatibility check off too.

Like I said, there is *not* an easy way to fetch that. Unless you mean from existing characters in the world.
In which case enter admin mode (`/admin`)
Hover your mouse over an npc
Type `/entityeval npc.species()` into chat
Look at the result in the log

Alternatively, if you are in singleplayer and want to skip the log, try
`/entityeval world.sendEntityMessage(-65535, "queueRadioMessage", {messageId="debug",unique=false,text=npc.species()})`

**Reply by ErinaSugino** [reply]
1. There is support for infinite-actor anims. Well, until your PC explodes. And you need to create all the artwork and positions and patch the already 10k lines long animation to, like, triple it's size. Hence nobody ever made more than threeactors, for only 4/8 positions - because they could reuse sprites from other positions.
2. Settings changed in-game could only be changed for the current session, and only for the current world. Starbound has literally 0 means to communicate data outside of a world instance. And changing global settings per player is never good, will just break stuff.
3. If you at that point already found that AI, you could just as well let it generate it for you and make your own addon. But assuming it's a good one, it'll be paid. And just as you don't pay anything to play this mod, I don't pay anything to develop it's content.
4. That sounds more like a bug than a feature request. Because that's how it should be anyway.
5. That's something that already crossed my mind, for later.
6. If you tell me how to *dynamically* patch a dungeon with possible spawns for custom races that *may or may not exist*, sure.

**Reply by ErinaSugino** [reply]
Looking through the dungeon json file, in it's current form the list of races for the spawning NPCs are not patchable. They are a string list, not a list list, so you can't safely append values, just completely override them. I also see some data defining that that property is a "string", so maybe one can change that, but I have literally 0 experience in working with dungeons, especially so with non-vanilla stuff. I also remember there being a discussion with some coding with spawner items required in the Neki mod to add an npctype as a chance spawn to a vanilla dungeon, so idk.

Also I only said I'm not writing dialogue myself. Not my field of work, as proven in the past. What others do is none of my concern, unless they make it one.

Ways to support my work can be found in the "Sponsor" section at the main page ("Code" tab).

**Reply by ErinaSugino** [reply]
Wow, AI is scary. Should ask ChatGPT "Can you fix Starbound?"

But again, primary course of action for adding *custom race specific data*, like dialogue or sprites or stuff, should first and foremost go into their own addon. People don'T need to download tons of additional data for races they don'T even play with, instead they can opt to install the addon when they want it.

Alternatively, if you are convinced people would benefit from having it SBR itself, make a pull request and I'll judge it. I know the already included "Extended Dialogues" mod adds alot of dialogue for other races, however that one was merge because it's long since abandoned and incompatible with SBR, so I just inhaled it to save everyone the trouble. And fix some stuff while at it.

<hr>

The "arousal of 100" cannot "not be satisfied" for NPCs, as they immediately run to the next free sexnode, and once they leave it by any means it's back to 0. The heat mechanic is a different topic and should be discussed and contributed to in it's respective thread, to keep things easily managable.

**Reply by ErinaSugino** [reply]
Don't understimate tables. But ok, assuming an NPC gets horny but nothing is around for miles - then they are supposed to get a debuff instead? Of what kind, what purpose? What does it do?

Arousal is a resource that's mainly used as a switch. If an NPC's arousal reaches >=100, their behaviour switches to "find sexnode". And once they enter it, they're waiting. Leaving the sexnode (on their own accord) happens only by two conditions - either they wait all alone for eternity and the 5 minute timeout kicks in (or rather, kicks *them out*) or they have sex, and through the arousal penalty of multiple climaxes they reach 0 arousal and leave. Independent of this, however, if they ever leave the sexnode by any means their arousal is forcibly set to 0. To ensure that "switch" is flipped back after they're "done". Imagine an NPCs having 100 arousal, waiting in a node, aborting after 5 minutes of waiting just to *immediately get back in* because arousal is still 100 and their AI says "horny, node".

<hr>

1. Once masturbation solo actor positions are a thing. No need before that.
2. That already exists.
3. Depends. Are we solely talking about what gushes out as an actual liquid in the world? Yes. Lewdbound already has a wide variety "racial sexual fluids" to collect afaik, would just need to expand SBRs config for ejaculation liquid to be a nested list for different races, then someone can patch those things together. Are we talking differently colored splurting particles in the animation? Hell no. Because the Starbound animator is so incredibly fucked up static, everything has to be pre-defined in the animation file. Which means, for every variation of climax particles there needs to be an emitter defined in the scene. Which means, currently that is 2 different genders per actor per position. Which then totals to the current number of 44 different defined particle emitters, just for climax particles. And this number only goes up with new positions. Now imagine you want to have differently colored particles for every race, obviously only counting the 8 vanilla races: **352 particle emitters**. Loaded at the same time, in every sexnode individually. Even though 95% of those are not used at all. Seems a bit too much effort for the result, and is certainly not doing good things to the framerate or RAM usage especially when having, like, an entire house full of fuckable goodies on the screen.
4. If there was a surefire, safe, not framerate nuking way to change ANYTHING visually on the actualy entity, outside of the sexnode, I would have jumped on that opportunity years ago. Sadly, there isn't. Mods like Futura's dragon engine "solve" the problem by just hiding the actual entity (like sexbound, causing incompatibilities) and then doing they ENTIRE render process themselves, *again*, in Lua. Performance goodbye. Mods like StarExtension give options to just edit the player entity like we should be able to anyway, but only the player, because modding the server side has been impossible for the longest part. Options are there - none of them ideal.
5. If at all low priority
6. Just add whatever music you want ¯\_(ツ)_/¯ Slap those sound files (.mp3 or .ogg I think) into the music folder, add them to the list in the config, profit.
7. Reasonably speaking, there would never be enough liquid to even amass on the ground to be collected anyway. Best bet would be making an object like Lustbound's milking machine, but for breasts (forgot the actual term) that produces that item when you're lactating.
8. I am adding an in-between step to "fully grown ass adult" and "pregnancy status effect" to 1.1, to add more immersion and easier discernability to "kids", but I'm not gonna implement a whole ass parent simulator.
9. I... am actually currently unsure if the status effect is hidden on maximum immersion mode or not. If not, the problem would be that you can't change the icon of a status effect at runtime to my current knowledge, so I would probably have to create two identical effects, one with the icon, one without, and juggle both at the same time to achieve that.

**Reply by ErinaSugino** [reply]
"That shouldn't be too hard to add".

Well, yes. If the experience you are looking for is Rimworld's adult mod, I think conveniently named "Rimjobs" or something? Where your entire colony dies week one because some couldn't keep their libido in check and goodo old Randy was like "Hey, that's pretty good!"

Implementing a couple status effects, copy and paste the syringe items from vanilla Starbound for "cures" and applying that shit on random every time you do it is simple. Properly integrating that feature in a meaningful way and, most importantly, *balancing it properly* is a masters degree study. E.g., let's talk about the cures. You suggested this feature because of "harsh punishing". Hence these "status effects" are "infinite" (status effects can't be, they are gone when you die. Should they be? Logically there is no reason why the clone pod would recreate you with deseases. But then again, where is the harsh punishment when you can just yeet yourself off the next cliff to solve all your problems?). Then how do you get the cures? Buy them for money? Great. Whole feature reduced to just running over to the outpost for 30 seconds and buying a dose everytime you fuck. Very punishing - in the overall enjoyability of gameplay. Make them expensive af? Money never really was a problem in late-game Starbound, and people tend to just cheat everything in they deem too tedious. So, same was with the potential option of "Buy for WhatEverTheEndGameTier7DungeonCurrencyWasCalled?", the answer would be (Cheater excluded) "Welcome to Sexbound, please don't touch anything until you beat the game or you will regret it." Buy it with Sexbux? Great. In order to heal an SDT, you need to fuck with an SDT. Yep. Find them as loot in the world? The cycle of balancing questions repeats. Which loot, where, how rare? Also, despite it ultimately being my goal od Sexbound being *an addition to* Starbound, not a replacement for horny people, most players see and use the mod as this. Being forced to explore endlessly repetetive worlds and dungeons to cure your itchy crotch goes very much against their interests.

How and how often do they trigger? Random is easy - but lackluster. Control is the keyword. If you just have to deal with a random chance it feels frustrating. If it follows a logical pattern that you understand and control, it's engaging. SDTs usually don't come out of nowhere. They are per person. So that would require a shitton of extra genetical data trying to gauge how likely a new NPC has one of those, so that you can then get it yourself fucking this specific NPC, if not careful. Speaking of careful, the only IRL known safe way to mostly remove the risk of SDTs is using a condom. Well, have fun female players, as there is no way to tell NPCs to use items. They don't even logically know the concept of usable items in vanilla Starbound.

Tl;dr NO

**Reply by ErinaSugino** [reply]
"The game doesn't really punish you in any way possible right now"
It is a video game. It cannot ever truly be punishing if player's aren't willing to take it serious. IRL pregnancies are a life-changing thing, one way or another. In Starbound you can give 0 fucks, duke the born NPC in lava and be on your merry way. Actual, globally applicable punishment would only involve the player's ability to play in the first place. Do what ETS2 for example does and just automatically remove money or stuff from their inventory everytime they violate the rules. Or teleport them to the jail dimension. Doesn't however mean the concept isn't solid. Pregnancy mechanics, opposite to your "simple" STD mechanics, are complex, but still understandable and controllable in many ways. Both if you want to or don't want to get pregnant, you can just leave it up to chance and do it. Or you can control your odds for one outcome or another. Use protection. Use a fertilitypill (we don't talk about the debug testing item of the hyperfertilitypill here). Time it with the ovulation cycle. Choose your mate wisely. So many options you can choose to influence the outcome in your favor, while still most of the time baring a risk of failure.

And that's why simply slapping out STDs left right and center, however RNGesus feels like it, is too simple a mechanic. No control (other than always wearing a contraceptive patch or just not fucking at all), no fun. In a real world scenario, taking control is simple: Step 1, ask your partner if they have anything to look out for. For most people, the answer is no and the instructions end here. If the answer is yes, then it goes about waging the risks and choosing the means of prevention. Now here is the funny thing: The first already is impossible in Starbound. You can't talk to an NPC like you can to another human. So that would require to develop any means to even query this important information beforehand. And then use it. An item that can be "used on" an entity and tells you if they have STDs? Maybe. Integrate it into the SexUI? Maybe already too late. What do NPCs do with this information? Update their AI to just never fuck with someone who has STDs? Or do we simplyfy NPCs in this equation, like with most mechanics, as they don't profit from immersive stuff and they are just fully 100% unaffected by all of this? So that only players can actually get and maybe spread the STD debuffs?

As for the part where you don't want me to scream at you: Yes, that would be a good draft on how to distribute chances. Which then however means were are already long since away from the most simplistic approach at the beginning, and at individually generating traits for every NPCs in the game. And I haven't even done that yet for subgenders in the wild.

And obviously this would be in the settings, with the first one being a big fat "off" switch. Because otherwise, a *lot* of people would be screaming at *me*.

But here's the thing. Since you're so dead set on this being a good addition, and keeping it simple, why don't you make it yourself, either as an addon or pull request here on GitHub? In it's most simplistic (and flawed) form, 95% of the work is just copy'n'pasting items and status effects, with only the last 5% of work involving code. I gotta focus on 1.1, and this certainly isn't 1.1 stuff.

**Comment by Mixarar** [comment]


**Reply by ErinaSugino** [reply]
1. Since that is already a general plan, I'll think up the proper way of implementation when the time comes.
2. That, as well as NPC sub-genders in the wild, is already planned.
3. ...what?
4. That's an idea I too already had. Only problem now is, as nicely pseudo-realistic as it is, it kinda clashes with the new "increased fertility upon successive climaxes" thingie. So that's kinda problematic.
5. See below
6. If Starbound had an easy way to give items to NPCs, the Starbound modding world would be a much cooler place. Alas, NPCs don't even know the concept of items or an inventory. They can have armor and a weapon. So that would require a loooot of workaround hacking.
7. It's a setting. When disallowed, NPCs refuse to go to a node with a family member. The player can still of course transform them into a node on the spot and then fuck their own family, can't prevent that (or rather, do not want to go that far). In either way, babies with family members will have a much higher stillborn chance, yes.
8. Impossible - unless I clutter entity storage data, and even then it's unreliable. NPCs only remember the UUID of their parents. Querying the corresponding entity one level deep would already only work if both the father and mother still existed, and are currently loaded in the same world and same location. Much less likely to succeed multiple levels deep. In order for a proper family tree, upon generating a baby the entire portrait data of the parents needs to be stored, and every NPC would need to store the data for the entire tree - at which point I'd hard-lock family at 3 generations simply for data size's sake.
9. Reworking that kinda stuff is on the agenda already, but doing something that works, makes sense AND is fun to use is hard.
Example: females slower, males faster. In a normal NPC x NPC setting as it currently stands, they'd never climax together again. Once again making it impossible to ever account for NPC x NPC scripted climax. When speaking of more realistic climax gain, you'd also have to think about how one cannot infinitely hold their climax back. Then we'd be at the point where you'd need detailed control over climax gain (like faster, slower, different positions etc.) to have a chance to ever cum together...

**Comment by XspeedPL** [comment]
Instead of/in addition to arousal controlling whether the scene can continue, maybe energy or hunger stats could be used?
The stat could be consumed continuously by taking an active role in a scene and also a bigger amount by climaxing.

Then, a variety of enhancer consumable items could be introduced that reduce the stat drain rates.

On a side note, maybe pregnancy should double the hunger rate at all times?

**Reply by ErinaSugino** [reply]
That would require to first teach NPCs the concept of hunger. Technically not too bad, just another resource they keep track of that automatically deltas like arousal, but you always have to keep in mind that NPCs are barebones robots that act like humanoids. Most common features of player characters have no meaning to them.

As for said players though - increasing the loss of food during pregnancy is actually a cool idea. Will keep that in mind.

**Reply by XspeedPL** [reply]
Right, I should have made it clear that I meant players only, since right now it's players that can keep scenes going indefinitely

**Reply by ErinaSugino** [reply]
Ah, ok. Since you said "in addition to arousal" I was assuming you meant NPCs, as players - as you now rightfully clarified - are not limited by anything at the moment.

In that case, I do very much like that idea. Which then however ties back into the perpetually ongoing discussion of #9 of how to utilize arousal for players in general. It makes sense that arousal should affect players in some way. But should it also be the determining factor for if they are *able to have sex in the first place?* Implementation sided that makes it easier. Logically, under most circumstances you would be able to keep going (especially as a female) even when not horny anymore. So another mechanic? Tying it to food seems a little drastic - and too non-limiting. The average player probably plays on Casual difficulty without food anyway, and this would also create kind of an infinite loop with fellatio (as there is the easter egg that fellatio climax increases food levels).

So this needs some serious thought.
